statix.solver/src/main/java/mb/statix/constraints/CCompiledQuery.java

```diff
@@ -108,11 +149,12 @@ public StateMachine<ITerm> stateMachine() {
                 subst.apply(resultTerm),
                 cause,
                 message == null ? null : message.apply(subst),
+                origin == null && trackOrigin ? this : origin,
                 stateMachine
         );
     }
 
-    @Override public CCompiledQuery unsafeApply(Immutable subst) {
+    @Override public CCompiledQuery unsafeApply(ISubstitution.Immutable subst, boolean trackOrigin) {
         return new CCompiledQuery(
                 filter.unsafeApply(subst),
```

Comment:
If you want to trace & compare dataWf's etc, `trackOrigin` needs to be passed here, down to the underlying rule body, as well.

Reply:
And similar to other occurrences in this file.

Reply:
I added code to track origins into apply() on the underlying rules.